pyflakes/checker.py

```diff
@@ -518,6 +534,60 @@ def is_typing_overload_decorator(node):
     )
 
 
+def make_tokens(code):
+    # PY3: tokenize.tokenize requires readline of bytes
+    if not isinstance(code, bytes):
+        code = code.encode('UTF-8')
+    lines = iter(code.splitlines(True))
+    # next(lines, b'') is to prevent an error in pypy3
+    return tuple(tokenize_tokenize(lambda: next(lines, b'')))
+
+
+class _TypeableVisitor(ast.NodeVisitor):
+    """Collect the line number and nodes which are deemed typeable by
+    PEP 484
+
+    https://www.python.org/dev/peps/pep-0484/#type-comments
+    """
+    def __init__(self):
+        self.typeable_lines = []  # type: List[int]
+        self.typeable_nodes = {}  # type: Dict[int, ast.AST]
+
+    def _typeable(self, node):
+        # if there is more than one typeable thing on a line last one wins
+        self.typeable_lines.append(node.lineno)
+        self.typeable_nodes[node.lineno] = node
+
+        self.generic_visit(node)
+
+    visit_Assign = visit_For = visit_FunctionDef = visit_With = _typeable
+    visit_AsyncFor = visit_AsyncFunctionDef = visit_AsyncWith = _typeable
+
+
+def _collect_type_comments(tree, tokens):
+    visitor = _TypeableVisitor()
+    visitor.visit(tree)
+
+    type_comments = collections.defaultdict(list)
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not TYPE_COMMENT_RE.match(text) or  # skip non-type comments
+                TYPE_IGNORE_RE.match(text)  # skip ignores
+        ):
+            continue
+
+        lineno, _ = start
+        idx = bisect.bisect_right(visitor.typeable_lines, lineno)
```

Comment:
my goal with "associate to previous" was to get the proper scope in which it has an effect such that pyflakes can check the right variables there.  you're right this example doesn't quite make sense and that `typed-ast` errors on this.  I couldn't quite make this happen with just cross referencing.

The actual reason it's "look backwards" is for functions which are annotated as follows:

```python
def f(
    a,  # type: int  # (annotation for a specific argument)
):
    # type: (...) -> None # (annotation for the function's return value)
    pass
```

In this case it's necessary to assign both the arguments and the type signature to the function itself, and make sure to evaluate those in the module scope instead of the function scope

Reply:
I'm adding a test and comment comment demonstrating and explaining this limitation.